### DISCUSSION TITLE
How does the SmptCommandException work when there are multiple recipients not accepted ?

### DISCUSSION THREAD

**Discussion by gbrunton** [jstedfast/MailKit#1751] [Q&A] [answered]
Can you help me understand what will happen in the following use cases when the server recognizes that an email address is not valid because the recipient is not accepted.

**Use Case 1**
```
mimeMessage.To.AddRange(new []
{
	new MailboxAddress("valid", validAddress1),
	new MailboxAddress("not-valid", notValidAddress), // this represents a recipient that is not accepted
	new MailboxAddress("valid", validAddress2)
});
smtpClient.Connect(....);
smtpClient.Send(mimeMessage);
```
I believe a SmptCommandException will be thrown. If that's true, should I assume that the valid addresses did not get the email?

**Use Case 2**

```
mimeMessage.To.AddRange(new []
{
	new MailboxAddress("not-valid", notValidAddress1), // this represents a recipient that is not accepted
	new MailboxAddress("not-valid", notValidAddress2) // this represents a recipient that is not accepted
});
smtpClient.Connect(....);
smtpClient.Send(mimeMessage);
```
I believe an exception will be thrown but if it's a SmptCommandException ex, then will the ex.Mailbox.Address equal the notValidAddress1 value and there won't be any reference to the notValidAddress2 value?

In this [comment](https://github.com/jstedfast/MailKit/issues/185#issuecomment-101235867), @jstedfast states that an exception gets thrown for the first recipient that fails but does not mention if valid email addresses will still get the email. His response also implies that only the first failed address is referenced in the exception. This makes me think that no further processing is done but I'm not sure.

**Comment by jstedfast** [comment, marked as answer]
You're asking a more complicated question than you realize because it "all depends" (on the server).

#### Use Case 1

> I believe a SmptCommandException will be thrown. If that's true, should I assume that the valid addresses did not get the email?

Assuming that the server rejects the email address outright, then yes. If it doesn't, then no. Instead, what will happen is that you'll get a delivery failure email message to the sender account telling you it failed to deliver the message.

#### Use Case 2

> I believe an exception will be thrown but if it's a SmptCommandException ex, then will the ex.Mailbox.Address equal the notValidAddress1 value and there won't be any reference to the notValidAddress2 value?

Assuming that the server rejects the email address outright, then yes. If it doesn't, then no. Instead, what will happen is that you'll get a delivery failure email message for each of the invalid recipients to the sender account telling you it failed to deliver the messages.

> In this https://github.com/jstedfast/MailKit/issues/185#issuecomment-101235867, @jstedfast states that an exception gets thrown for the first recipient that fails but does not mention if valid email addresses will still get the email. His response also implies that only the first failed address is referenced in the exception. This makes me think that no further processing is done but I'm not sure.

The way that the SMTP protocol works is that the client sends multiple commands to send a message.

* First, it will send a `MAIL FROM` command with the sender email address. If that fails, no further processing happens.
* Then a `RCPT TO` command is sent for each recipient address. As soon as any of them fail, the command sequence is terminated and an exception is thrown for the failed recipient address.
* If the `MAIL FROM` and `RCPT TO` commands all succeed, then the message itself is sent in a `DATA` command.

That should clear a lot of things up for you.

Keep in mind that the server won't necessarily do a lookup of any of the recipient addresses (but it might check that the domains exist) during the `RCPT TO` commands, so it won't know if the recipient addresses exist *unless* the recipient address is managed by the same host as the SMTP server (then it will usually know right away).
